src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,167 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] array = (T[]) new Object[DEFAULT_CAPACITY];
+    private int listLength;
+
     @Override
     public void add(T value) {
-
+        if (listLength < array.length) {
+            array[listLength] = value;
+            listLength++;
+        } else {
+            array = grow(array);
+            array[listLength] = value;
+            listLength++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + listLength);
+        }
+        if (index < array.length && index < listLength) {
+            if (listLength == array.length) {
+                array = grow(array);
+            }
+            for (int i = array.length - 1; i > index; i--) {
+                array[i] = array[i - 1];
+            }
+            array[index] = value;
+            listLength++;
+        } else if (index == listLength) {
+            this.add(value);
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + listLength);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (list.size() + listLength > array.length) {
+            array = grow(array);
+        }
+        if (listLength == 0) {
+            for (int i = 0; i < list.size(); i++) {
+                array[i] = list.get(i);
+                listLength++;
+            }
+        } else {
+            int shift = indexOf(null);
```

Comment:
The 'indexOf(null)' method seems to be used to find the first null element, which is not a reliable way to handle the addition of all elements. This approach could lead to unexpected behavior if null elements are valid in the list. Consider using 'listLength' directly for the starting index of the copy.